src/frontend/python/chiquito/cb.py

```diff
@@ -222,3 +227,93 @@ def to_constraint(v: ToConstraint) -> Constraint:
         raise TypeError(
             f"Type `{type(v)}` is not ToConstraint (one of Constraint, Expr, int, or F)."
         )
+
+
+@dataclass
+class LookupTableRegistry:
+    tables: Dict[int, LookupTable] = field(default_factory=dict)
+
+    def add(self: LookupTableRegistry, table: LookupTable):
+        self.tables[table.uuid] = table
+
+    def get(self: LookupTableRegistry, uuid: int) -> LookupTable:
+        return self.tables[uuid]
+
+
+@dataclass
+class LookupTable:
+    uuid: int = 0
+    dest: List[Expr] = field(default_factory=list)
+
+    def __init__(self: LookupTable):
+        self.uuid: int = uuid()
+        self.dest = []
+
+    def add(self: LookupTable, expr: ToExpr) -> LookupTable:
+        self.dest.append(to_expr(expr))
+        return self
+
+    def apply(self: LookupTable, constraint: ToConstraint) -> LookupTableBuilder:
+        return LookupTableBuilder(self.uuid).apply(constraint)
+
+    def when(self: LookupTable, enable: ToConstraint) -> LookupTableBuilder:
+        return LookupTableBuilder(self.uuid).when(enable)
+
+
+@dataclass
+class LookupTableBuilder:
+    uuid: int
+    src: List[Constraint] = field(default_factory=list)
+    enable: Optional[Constraint] = None
+
+    def __init__(self: LookupTableBuilder, uuid: int):
+        self.uuid: int = uuid
+
+    def apply(self: LookupTableBuilder, constraint: ToConstraint) -> LookupTableBuilder:
+        self.src.append(to_constraint(constraint))
+        return self
+
+    def when(self: LookupTableBuilder, enable: ToConstraint) -> LookupTableBuilder:
+        if self.enable is not None:
+            raise ValueError("LookupTableBuilder: when() can only be called once.")
+        self.enable = to_constraint(enable)
+        return self
+
+    def build(self: LookupTableBuilder, step_type: StepType) -> Lookup:
+        table = step_type.tables.get(self.id)
```

Comment:
In Rust Chiquito, table is initially built using `new_table` from DSL and stored in `CircuitContext`. It's then cloned to `StepTypeContext` and `StepTypeSetupContext`. Because `build` is called by `add_lookup`, a `StepTypeSetupContext` function, `build` conveniently takes a `ctx: StepTypeSetupContext` parameter. Therefore, table doesn't "have to be" stored in the step, but it's just the most convenient for `build` to have the `step_type` parameter, as `add_lookup` calls `build` at the step type level.

In Python, `add_lookup` also calls `build` at the step type level. The tables are initially created and stored at the circuit level, but step type and super circuit can also have references to the tables.